File: model/flashcard.py
import json
from model.type import *


class Flashcard:

	# Loads a flashcard from a json file.
	def __init__(self, path):
		self._path = path
		self._init = False
		try:
			with open(path, "r", encoding="utf-8") as file:
				data = json.load(file)

				# Gets the card type.
				self._type = data.get("type", None)
				if type(self._type) != str:
					print(f"[SESHAT]: error: '{path}' must define 'type' string field")
					return
				self._type = CardType.from_id(self._type)
				if self._type == None:
					print(f"[SESHAT]: error: 'type' in '{path}' must be word, kanji or radical")
					return

				# Loads the corresponding card.
				match self._type:
					case CardType.WORD:		
						if not self._load_word(data) : return
					case CardType.KANJI:
						if not self._load_kanji(data) : return
					case CardType.RADICAL:	
						if not self._load_radical(data) : return

		# Error managment.
		except FileNotFoundError:
			print(f"[SESHAT]: error: unable to open '{path}'")
			return
		except json.JSONDecodeError:
			print(f"[SESHAT]: error: '{path}' isn't a valid json file")
			return
		self._init = True

	# Loads a word flashcard.
	def _load_word(self, data):
		if not self._load_common(data) : return False
		self._reading = data.get("reading", None)
		self._word_class = data.get("class", None)
		if type(self._reading) != str: 
			print(f"[SESHAT]: error: '{self._path}' must define a 'reading' string field")
			return False
		if type(self._word_class) != str: 
			print(f"[SESHAT]: error: '{self._path}' must define a 'class' string field")
			return False
		self._word_class = WordType.from_id(self._word_class)
		if self._word_class == None:
			print(f"[SESHAT]: error: '{self._path}' must define a valid 'class' field")
			return False
		return True
	
	# Loads a kanji flashcard.
	def _load_kanji(self, data):
		if not self._load_common(data) : return False
		self._on = data.get("on", None)
		self._kun = data.get("kun", None)
		self._radical = data.get("radical", None)
		if type(self._on) != str:
			print(f"[SESHAT]: error: '{self._path}' must define a 'on' string field")
			return False
		if type(self._kun) != str:
			print(f"[SESHAT]: error: '{self._path}' must define a 'kun' string field")
			return False
		if type(self._radical) != list:
			print(f"[SESHAT]: error: '{self._path}' must define a 'radical' list field")
			return False
		for r in self._radical:
			if type(r) != str: 
				print(f"[SESHAT]: error: '{self._path}' 'radical' field must contain only strings")
				return False
		return True

	# Loads a radical flashcard.
	def _load_radical(self, data):
		if not self._load_common(data) : return False
		self._name = data.get("name", None)
		if type(self._name) != str:
			print(f"[SESHAT]: error: '{self._path}' must define a 'name' string field")
			return False
		return True

	# Loads the common flashcard data.
	def _load_common(self, data):
		self._front = data.get("front", None)
		self._meaning = data.get("meaning", None)
		self._mnemonic = data.get("mnemonic", None)
		self._example = data.get("example", None)
		if type(self._front) != str: 
			print(f"[SESHAT]: error: '{self._path}' must define a 'front' string field")
			return False
		if type(self._meaning) != str: 
			print(f"[SESHAT]: error: '{self._path}' must define a 'meaning' string field")
			return False
		if type(self._mnemonic) != str: 
			print(f"[SESHAT]: error: '{self._path}' must define a 'mnemonic' string field")
			return False
		if type(self._example) != list: 
			print(f"[SESHAT]: error: '{self._path}' must define an 'example' list field")
			return False
		for e in self._example:
			if type(e) != str: 
				print(f"[SESHAT]: error: '{self._path}' 'example' field must contain only strings")
				return False
		return True

	# Gets the initialisation state of the flashcard.
	@property
	def init(self):
		return self._init

	# Gets the type of the flashcard.
	@property
	def type(self):
		return self._type
	
	# Gets the front string of the flashcard.
	@property
	def front(self):
		return self._front
	
	# Gets the meaning string of the flashcard.
	@property
	def meaning(self):
		return self._meaning
	
	# Gets the mnemonic string of the flashcard.
	@property
	def mnemonic(self):
		return self._mnemonic
	
	# Gets the example list of the flashcard.
	@property
	def example(self):
		return self._example

	# Gets the name string of the flashcard in case of a radical.
	@property
	def name(self):
		if self._type != CardType.RADICAL: return None
		return self._name

	# Gets the on string of the flashcard in case of a kanji.
	@property
	def on(self):
		if self._type != CardType.KANJI: return None
		return self._on

	# Gets the kun string of the flashcard in case of a kanji.
	@property
	def kun(self):
		if self._type != CardType.KANJI: return None
		return self._kun

	# Gets the radical list of the flashcard in case of a kanji.
	@property
	def radical(self):
		if self._type != CardType.KANJI: return None
		return self._radical

	# Gets the reading string of the flashcard in case of a word.
	@property
	def reading(self):
		if self._type != CardType.WORD: return None
		return self._reading
	
	# Gets the word class string of the flashcard in case of a word.
	@property
	def word_class(self):
		if self._type != CardType.WORD: return None
		return self._word_class

File: static/js/flashcard.js
allowFlip = false;
function cardsAllowFlip() {allowFlip = true;}
function cardsPreventFlip() {allowFlip = false;}


// Flashcard base class.
class Flashcard {
	constructor(type, front, meaning, mnemonic, example) {
		this.type = type;
		this.front = front;
		this.meaning = meaning;
		this.mnemonic = mnemonic;
		this.example = example;
	}

	// Must be overriden by the children.
	render(index) {
        throw new Error("render() wasn't implemented");
    }

	// This renders the shared first part of the card HTML script.
	renderPre(index) {
		return `
			<div class="flashcard ${this.type} ${index === 0 ? 'active' : ''}" 
			${allowFlip ? 'onclick="this.classList.toggle(\'flipped\')"' : ''}>
				<div class="flashcard-inner">
					<div class="flashcard-face flashcard-front">
						${this.front}
					</div>
					<div class="flashcard-face flashcard-back">
						<div class="card-content">
							<h3 class="card-title">${this.meaning.join(', ')}</h3>`;
	}

	// This renders the shared last part of the card HTML script.
	renderPost() {
		return `
							<p class="card-info"><strong>Mnémonique:</strong></p>
							<p>${this.mnemonic}</p>
											
							${this.example.length > 0 ? `
								<div class="examples">
									<p class="card-info"><strong>Exemples:</strong></p>
										${this.example.map(ex => `<p class="example">${ex}</p>`).join('')}
								</div>
							` : ''}
						</div>
					</div>
				</div>
			</div>`;
	}
}

// Flashcard class for words.
class WordCard extends Flashcard {
	constructor(type, front, meaning, mnemonic, example, data) {
		super(type, front, meaning, mnemonic, example);
		this.reading = data.reading;
		this.word_class = data.class;
	}

	// This function returns the HTML to render the card.
	render(index) {
		return `
			${this.renderPre(index)}
				<p class="card-info"><strong>Lecture:</strong> ${this.reading}</p>
				<p class="card-info"><strong>Type:</strong> ${this.word_class}</p>
			${this.renderPost()}
			`;
	}
}

// Flashcard class for radicals.
class RadicalCard extends Flashcard {
	constructor(type, front, meaning, mnemonic, example, data) {
		super(type, front, meaning, mnemonic, example);
		this.name = data.radical_reading;
	}

	// This function returns the HTML to render the card.
	render(index) {
		return `
			${this.renderPre(index)}
				<p class="card-info"><strong>Nom:</strong> ${this.name}</p>
			${this.renderPost()}
		`;
	}
}

// Flashcard class for kanji.
class KanjiCard extends Flashcard {
	constructor(type, front, meaning, mnemonic, example, data) {
		super(type, front, meaning, mnemonic, example);
		this.on = data.on;
		this.kun = data.kun;
		this.radical = data.radical;
	}

	// This function returns the HTML to render the card.
	render(index) {
		return `
			${this.renderPre(index)}
				<p class="card-info"><strong>On:</strong> ${this.on.join(', ')}</p>
				<p class="card-info"><strong>Kun:</strong> ${this.kun.join(', ')}</p>
				<p class="card-info"><strong>Radical:</strong> ${this.radical.join(', ')}</p>
			${this.renderPost()}
		`;
	}
}

// Loads a card from its json file.
async function loadCard(deck_name, card_data, type) {
	try {

		// Gets all the shared data between card types.
		const front = card_data.japanese_name ?? null;
		const meaning = card_data.meaning ?? null;
		const mnemonic = card_data.mnemonic ?? null;
		const example = card_data.example ?? null;

		// Creates the right card.
		switch (type) {
			case "word":
				return new WordCard(type, front, meaning, mnemonic, example, card_data);
			case "radical":
				return new RadicalCard(type, front, meaning, mnemonic, example, card_data);
			case "kanji":
				return new KanjiCard(type, front, meaning, mnemonic, example, card_data);
			default:
				throw new Error(`[SESHAT]: unknown card type '${type}'`);
		}

	// Error managment.
	} catch (error) {
		console.error(`Error when loading deck '${deck_name}' card :`, error);
		return null;
	}
}
